pkg/clustermesh/mcsapi/serviceexportsync.go

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package mcsapi
+
+import (
+	"context"
+	"sync"
+
+	"github.com/cilium/hive/cell"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	mcsapiv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	"github.com/cilium/cilium/pkg/clustermesh/mcsapi/types"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/kvstore/store"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "mcsapi")
+
+// ServiceExportResource builds the Resource[ServiceExport] object.
+func ServiceExportResource(lc cell.Lifecycle, cs client.Clientset, opts ...func(*metav1.ListOptions)) resource.Resource[*mcsapiv1alpha1.ServiceExport] {
+	if !cs.IsEnabled() {
+		return nil
+	}
+	lw := utils.ListerWatcherWithModifiers(
+		utils.ListerWatcherFromTyped(cs.MulticlusterV1alpha1().ServiceExports("")),
+		opts...,
+	)
+	return resource.New[*mcsapiv1alpha1.ServiceExport](lc, lw, resource.WithMetric("ServiceExport"))
+}
+
+type ServiceExportSyncParameters struct {
+	ClusterName    string
+	ServiceExports resource.Resource[*mcsapiv1alpha1.ServiceExport]
+	Services       resource.Resource[*slim_corev1.Service]
+	Backend        store.SyncStoreBackend
+	StoreFactory   store.Factory
+	Store          store.SyncStore
```

Comment:
I did this so that I can create the store from the test to directly have a kvstore client there instead of having to return it somehow but maybe you would have a better idea to do this :thinking:

Reply:
Fair enough. In that case, I'd unexport the `Store` field, and add a comment clarifying that it is used for testing purposes only. Another possibility could be modifying the test to not require mutating the SyncStore directly. Like first ensuring that the service export is present in the kvstore, then removing either the service or the service export and asserting that it gets removed. That would mimic even better the real behavior. Additionally, it would be better to use the fake client to create/remove the objects if possible, rather than mutating the underlying store, especially because it seems that the `CacheStore` method will be eventually dropped based on the comment.